### PR TITLE
don't include stdio.h in caml/misc.h

### PR DESCRIPTION
**Pull request by madroach**
On OpenBSD, `stderr` and `stdout` are macros defined in `stdio.h`
ppx_expect uses `stderr` and `stdout` as identifiers in
`collector/expect_test_collector_stubs.c` where `caml/misc.h` is included.
This confuses the C compiler, because the macro will get expanded where an identifier is expected.
There is no need to include `stdio.h` in `caml/misc.h`
Build error log of ppx_expect is [here](https://www.gmerlin.de/ppx_expect.txt)
This seems to have happened by accident in commit cddec18fde8

### CODE REVIEW

**Comment by xavierleroy** [comment]
I don't understand the issue with ppx_expect, but I don't want to know.

@madroach is correct that `<stdio.h>` is not used in `<caml/misc.h>` and that its inclusion is an accident (probably leftover debugging code).  Since it is a recent inclusion (4.10?), we can hope no 3rd-party code depends on it.  Pinging @jhjourdan for a second opinion.

(While we're at it we could also remove the fallback definition of `NULL`.)

**Review by xavierleroy — APPROVED** [review]
Looks good to me, as a revert of a probably unintended change.

**Comment by xavierleroy** [comment]
Amusingly (?), one of the tests in the test suite wrongly relied on `<stdio.h>` being included for free.
`/testsuite/tests/compatibility/stub.c` needs fixing.

**Comment by dra27** [comment]
Feels like it should be included in 4.10.1 as well

**Comment by dra27** [comment]
It's good to have found the error here, but ppx_expect should also be fixed - if my skimming of [GCC's](https://www.gnu.org/software/libc/manual/html_node/Reserved-Names.html) and [Posix/ISO C's](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html) rules are correct, `stdout` and `stderr` should not be used for variable names, regardless of whether `stdio.h` has actually been included.

**Comment by yallop** [comment]
Using `stdout` as anything other than an expression if `<stdio.h>` is included isn't valid, and defining a global variable called `stdout` is invalid in any case.  But defining a local variable called `stdout`, as `ppx_expect` does, is valid AFAICS (if arguably unwise).

**Comment by xavierleroy** [comment]
Cherry-picked to 4.11 in commit 8cf2ecd15

**Comment by dra27** [comment]
4.10 too?

**Comment by xavierleroy** [comment]
> 4.10 too?

Morally, yes.  Technically, it is a breaking change for code that just works in 4.10.0.

**Comment by gasche** [comment]
On the other hand, the faulty include was first released in 4.10, so it means that there is just one version that makes this code work. I think it would be preferable to have this in a 4.10 bugfix release, so that we have the nice no-surprise property that such code works or not in the same way with all  (latest bugfix releases of) major versions.

**Comment by xavierleroy** [comment]
This will be for the 4.10.1 release manager to decide. (@Octachron)

**Comment by Octachron** [comment]
Fixing an accidental header inclusion in 4.10.1 seems better to me to maintain coherence along all latest version. I will run a check on the opam repository as a precaution however.

**Comment by avsm** [comment]
@engil points out in a private thread that #9082's rebase brings stdio.h back in, since it uses `FILE *` in the domain_state information.  Fixing ppx_expect just seems the easiest approach here, as (although it is very technically valid), the use of the `stdout/stderr` names seems very unwise in portable C code.

**Pull request merged**
